File: Feb-2023/Java/BinaryTreeZigzagLevelOrderTraversal.java
// 103. Binary Tree Zigzag Level Order Traversal

// Keep a toggle boolean value, to reverse the elements while transversing for making a zigzag pattern

/**

 * Definition for a binary tree node.

 * public class TreeNode {

 *     int val;

 *     TreeNode left;

 *     TreeNode right;

 *     TreeNode() {}

 *     TreeNode(int val) { this.val = val; }

 *     TreeNode(int val, TreeNode left, TreeNode right) {

 *         this.val = val;

 *         this.left = left;

 *         this.right = right;

 *     }

 * }

 */

 

class Solution {

    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {

        if(root==null) return new ArrayList<>();

        Queue<TreeNode> q = new LinkedList<>(); 

        List<List<Integer>> res = new ArrayList<>();

        q.add(root);

        boolean reversecheck=false;

        while(!q.isEmpty()){           

            int size = q.size();

            List<Integer> ls = new LinkedList<>();

            while(size>0){

                size--;

                root = q.poll();

                ls.add(root.val);

                if(root.left!=null) q.add(root.left);

                if(root.right!=null) q.add(root.right);                             }

            if(reversecheck) Collections.reverse(ls);

            res.add(ls);

            reversecheck = !reversecheck;

            }

        return res;

    }

}

File: Feb-2023/C++/BinaryTreeZigzagLevelOrderTraversal.cxx
// 103. Binary Tree Zigzag Level Order Traversal

// Keep a toggle boolean value, to reverse the elements while transversing for making a zigzag pattern

/**

 * Definition for a binary tree node.

 * struct TreeNode {

 *     int val;

 *     TreeNode *left;

 *     TreeNode *right;

 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}

 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}

 * };

 */

class Solution {

public:

    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {

        if(root==nullptr){

          return {};

        }

        bool reversecheck=true;

        vector<vector<int>> res;

        queue<TreeNode*> q;

        q.push(root);

        while(!q.empty()){

            int size = q.size();

            vector<int> ls;

            while(size-->0){

                TreeNode* x = q.front();

                q.pop();

                ls.push_back(x->val);

                if(x->left!=nullptr){

                    q.push(x->left);

                }

                if(x->right!=nullptr){

                    q.push(x->right); 

                }            

            }

            if(!reversecheck){

               reverse(ls.begin(),ls.end());

             }

                res.push_back(ls);

                reversecheck=!reversecheck;

        }

        return res;

    }

};
